pkg/core/typeurl/types.go

```diff
@@ -0,0 +1,143 @@
+/*
+Copyright 2019 The KubeEdge Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package typeurl is Marushal For any type
+package typeurl
+
+import (
+	"encoding/json"
+	"path"
+	"reflect"
+	"sync"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/gogo/protobuf/types"
+	"github.com/pkg/errors"
+)
+
+var registry map[reflect.Type]string
+var revRegistry map[string]reflect.Type
+var lock sync.Mutex
+
+var (
+	// WrongTypeError stand for custom type error
+	WrongTypeError = errors.Errorf("Wrong Type Error")
+)
+
+func init() {
+	registry = make(map[reflect.Type]string)
+	revRegistry = make(map[string]reflect.Type)
+	lock = sync.Mutex{}
+	// Register sting
+	RegisterType("", "string.content")
+	// Register boolean
+	RegisterType(true, "bool.content")
+	// register slice
+	RegisterType([]interface{}{}, "slice.interface.content")
+}
+
+func tryStrip(v interface{}) reflect.Type {
+	t := reflect.TypeOf(v)
+	switch t.Kind() {
+	case reflect.Ptr:
+		// if kind is Prt, get it`s realy type
+		for t.Kind() == reflect.Ptr {
+			t = t.Elem()
+		}
+	case reflect.Chan, reflect.Func, reflect.Interface, reflect.UnsafePointer:
+		panic(WrongTypeError)
+	}
+	return t
+}
+
+// RegisterType is Register a type with the base url of the type
+func RegisterType(v interface{}, args ...string) {
+	t := tryStrip(v)
+	param := []string{"kubeedge.io"}
```

Comment:
can we define the base url as a constant ?

Reply:
ok